### ISSUE TITLE
panic: runtime error: slice bounds out of range [:2] with capacity 0

### ISSUE THREAD

**Issue by altergui** [vocdoni/vocdoni-node#1058]
```
2023-07-25T18:07:00.064Z DBG middleware/logger.go:158 > "GET https://api-dev.vocdoni.net/v2/elections/c5d2460186f756414cc0115a1c32ae3baf037ff5bd6390b95dbd020000000000 HTTP/2.0" from 181.208.114.90:46176 - 200 1438B in 1.000297ms
2023-07-25T18:07:00.065Z DBG middleware/logger.go:158 > "GET https://api-dev.vocdoni.net/v2/elections/c5d2460186f756414cc0115a1c32ae3baf037ff5bd6390b95dbd020000000000 HTTP/2.0" from 181.208.114.90:46176 - 200 1439B in 3.245205ms
2023-07-25T18:07:00.23Z DBG apirest/apirest.go:191 > request: {"txPayload":"CnsKeQogXAxR0MY18gshvgk2BaRWz5KGGjNvwOj/KDkoZyMsom0SIMXSRgGG91ZBTMARWhwyrjuvA3/1vWOQuV29AgAAAAAAGiQyIhogCgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAiDXsidm90ZXMiOlsxXX0SQYgQUtElNXs2LbTOmtLEVATMfykMfdqAKWSxtvnLs3ixfx28edbzOTV/0JclQP5vxgEz5SeW+WToRCeWJMqH3IoA"}
2023-07-25T18:07:00.233Z DBG transaction/vote_tx.go:183 > new vote address=0x0722e4820e44b16DadcD6D032fe0080604203635 electionID=c5d2460186f756414cc0115a1c32ae3baf037ff5bd6390b95dbd020000000000 height=2660 nullifier=8bd88ccd14999271840b76bacde7cdb7857942b407c01b480bc8be8800fb41c0 type=signature
2023-07-25T18:07:00.233Z DBG transaction/proof.go:47 > verify proof address=0722e4820e44b16dadcd6d032fe0080604203635 censusOrigin=2 censusRoot=5f910b78db6ab20e17661fa66054edc2a74c8ea2225fc3ab819d1db5d316c493 electionID=c5d2460186f756414cc0115a1c32ae3baf037ff5bd6390b95dbd020000000000 voterID=010390deb6fa29bcb8a428ffd141ea0415d5db493938a1d60fa86bae083ac2978e62
panic: runtime error: slice bounds out of range [:2] with capacity 0

goroutine 23502102 [running]:
go.vocdoni.io/dvote/tree/arbo.UnpackSiblings({0x3f94d10?, 0x57e5398?}, {0x0?, 0xc01629ce00?, 0x27b8e71?})
	go.vocdoni.io/dvote/tree/arbo/tree.go:1011 +0x617
go.vocdoni.io/dvote/tree/arbo.CheckProof({0x3f94d10, 0x57e5398}, {0xc01541f7a0, 0x14, 0x20}, {0xc01541f380, 0x20, 0x20}, {0xc01541f4c0, 0x20, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/tree.go:1101 +0x85
go.vocdoni.io/dvote/tree.VerifyProof(...)
	go.vocdoni.io/dvote/tree/tree.go:233
go.vocdoni.io/dvote/vochain/transaction.VerifyProofOffChainTree(0xa?, 0xa?, 0x176b57a0?, {0xc01541f4c0, 0x20, 0x20}, {0x19b4a3b?, 0x1f4?, 0x0?}, {0xc01448a8a0, ...})
	go.vocdoni.io/dvote/vochain/transaction/proof.go:115 +0x24e
go.vocdoni.io/dvote/vochain/transaction.VerifyProof(0xc005146380, 0xc01114e880, {0xc01448a8a0, 0x22, 0x30})
	go.vocdoni.io/dvote/vochain/transaction/proof.go:71 +0x45e
go.vocdoni.io/dvote/vochain/transaction.(*TransactionHandler).VoteTxCheck(0xc0036d7b00, 0xc0183a7340, 0x0)
	go.vocdoni.io/dvote/vochain/transaction/vote_tx.go:191 +0x11d5
go.vocdoni.io/dvote/vochain/transaction.(*TransactionHandler).CheckTx(0xc0036d7b00, 0xc0183a7340, 0x0)
	go.vocdoni.io/dvote/vochain/transaction/transaction.go:81 +0xde
go.vocdoni.io/dvote/vochain.(*BaseApplication).CheckTx(0xc003d16000, {0x0?, 0xc004a179a8?}, 0xc0164213a0)
	go.vocdoni.io/dvote/vochain/app.go:273 +0x1a8
github.com/cometbft/cometbft/abci/client.(*localClient).CheckTxAsync(0xc0047e8420, {0x3f95df8?, 0xc0000da008?}, 0xc0164213a0)
	github.com/cometbft/cometbft@v0.38.0-rc3/abci/client/local_client.go:51 +0xe7
github.com/cometbft/cometbft/proxy.(*appConnMempool).CheckTxAsync(0xc003718168, {0x3f95df8, 0xc0000da008}, 0x121896a?)
	github.com/cometbft/cometbft@v0.38.0-rc3/proxy/app_conn.go:147 +0x102
github.com/cometbft/cometbft/mempool.(*CListMempool).CheckTx(0xc0049fa0c0, {0xc00a85a000, 0xc0, 0xc0}, 0xc01114e700?, {0x7348?, {0x0?, 0x2710?}})
	github.com/cometbft/cometbft@v0.38.0-rc3/mempool/clist_mempool.go:268 +0x230
github.com/cometbft/cometbft/rpc/core.(*Environment).BroadcastTxSync(0xc003c42fc0, 0xc0046b7900, {0xc00a85a000, 0xc0, 0xc0})
	github.com/cometbft/cometbft@v0.38.0-rc3/rpc/core/mempool.go:35 +0xfd
github.com/cometbft/cometbft/rpc/client/local.(*Local).BroadcastTxSync(...)
	github.com/cometbft/cometbft@v0.38.0-rc3/rpc/client/local/local.go:98
go.vocdoni.io/dvote/vochain.(*BaseApplication).SetDefaultMethods.func4({0xc00a85a000?, 0xc0, 0x2710?})
	go.vocdoni.io/dvote/vochain/appsetup.go:66 +0x5a
go.vocdoni.io/dvote/vochain.(*BaseApplication).SendTx(0xc011000101?, {0xc00a85a000?, 0xc0002d3aa0?, 0x2710?})
	go.vocdoni.io/dvote/vochain/app.go:458 +0x33
go.vocdoni.io/dvote/api.(*API).sendTx(0xc00a85a000?, {0xc00a85a000?, 0xc0?, 0x3376de0?})
	go.vocdoni.io/dvote/api/helpers.go:47 +0x37
go.vocdoni.io/dvote/api.(*API).submitVoteHandler(0xc00c9f8ad0?, 0xc012fe2ff0, 0x0?)
	go.vocdoni.io/dvote/api/vote.go:71 +0xd8
go.vocdoni.io/dvote/httprouter/apirest.(*API).RegisterMethod.func1({{0x3220920, 0xc012fe2ff0}, {0xc12821b10dbb2e25, 0x156f42b4f3c, 0x57af4e0}, {0xc012fe3030, 0x2, 0x2}, 0xc016421380})
	go.vocdoni.io/dvote/httprouter/apirest/apirest.go:214 +0x74
created by go.vocdoni.io/dvote/httprouter.(*HTTProuter).routerHandler.func1
	go.vocdoni.io/dvote/httprouter/httprouter.go:261 +0x3f2
```

**Comment by mariajdab** [comment]
I identified and reproduced the error:

In the function `VerifyProofOffChainTree` 
```
	case *models.Proof_Arbo:
		p := proof.GetArbo()
		.
		.
		.
		valid, err := tree.VerifyProof(hashFunc, leafKey, p.LeafWeight, p.Siblings, censusRoot)
		if !valid || err != nil {
			return false, nil, err
		}
```
I think `VerifyProof` is called with a empty slice `[]byte{}` in `p.Siblings`

And  `VerifyProof` calls `CheckProof`
```
func CheckProof(hashFunc HashFunction, k, v, root, packedSiblings []byte) (bool, error) {
	siblings, err := UnpackSiblings(hashFunc, packedSiblings)
	if err != nil {
		return false, err
	}
```
```
func UnpackSiblings(hashFunc HashFunction, b []byte) ([][]byte, error) {
	fullLen := binary.LittleEndian.Uint16(b[0:2])
	l := binary.LittleEndian.Uint16(b[2:4]) // bitmap bytes length
	if len(b) != int(fullLen) {
		return nil,
			fmt.Errorf("expected len: %d, current len: %d",
				fullLen, len(b))
	}
```

And in `UnpackSiblings` is not being checked the len of b before chunk the slice.

So, I wonder is `p.Siblings` should be of a fixed size, so we can add a check validation in `VerifyProofOffChainTree` or the validation must be in the `UnpackSiblings` function?

**Comment by mariajdab** [comment]
@altergui @p4u

**Comment by altergui** [comment]
i'm not familiar with arbo, but i'd start UnpackSiblings with `if len(b) < 4 {return nil, fmt.Errorf() }`

then we can see if p.Siblings should be of a fixed size, but i wouldn't leave UnpackSiblings as it is now, in any case.

**Comment by p4u** [comment]
Great finding @mariajdab 
This is code currently being modified by @lucasmenendez So I'd ask him to include the fix to this issue within its pull request.

**Comment by lucasmenendez** [comment]
I think that the packed siblings has not a fixed size. As you can see in the code, the `UnpackSiblings` functions gets the first two bytes (that contains the length of the siblings bitmap) to compare the integer that this bytes contains with the length of the rest of the slice. So I think that, at least it will have 2 bytes and I suppose that something similar that the @altergui solution will work. I will include the fix of this in the anonymous PR and close this issue, thanks @mariajdab!

**Comment by mariajdab** [comment]
> I think that the packed siblings has not a fixed size. As you can see in the code, the `UnpackSiblings` functions gets the first two bytes (that contains the length of the siblings bitmap) to compare the integer that this bytes contains with the length of the rest of the slice. So I think that, at least it will have 2 bytes and I suppose that something similar that the @altergui solution will work. I will include the fix of this in the anonymous PR and close this issue, thanks @mariajdab!

Sounds Good @lucasmenendez, but I think that at least it will have 4 bytes due the code `l := binary.LittleEndian.Uint16(b[2:4])` that is in the same function `UnpackSiblings`

**Comment by lucasmenendez** [comment]
> Sounds Good @lucasmenendez, but I think that at least it will have 4 bytes due the code `l := binary.LittleEndian.Uint16(b[2:4])` that is in the same function `UnpackSiblings`

Yes @mariajdab, but I think that the declaration of `l` variable should be placed after the conditional that checks if the decoded size is equal to the size of the rest of the slice. So I propose a different approach:

```go
func UnpackSiblings(hashFunc HashFunction, b []byte) ([][]byte, error) {
	if len(b) < 2 {
		return nil, fmt.Errorf("no enought size of packed siblings provided")
	}

	fullLen := binary.LittleEndian.Uint16(b[0:2])
	if len(b) != int(fullLen) {
		return nil,
			fmt.Errorf("expected len: %d, current len: %d",
				fullLen, len(b))
	}

	l := binary.LittleEndian.Uint16(b[2:4]) // bitmap bytes length
	bitmapBytes := b[4 : 4+l]

        // ...
}
```

I also check that this works updating the test:

```go
func TestPackAndUnpackSiblings(t *testing.T) {
        // ...

        // test edge cases with invalid packed siblings
        _, err = UnpackSiblings(HashFunctionPoseidon, []byte{})
        c.Assert(err, qt.IsNotNil)
        _, err = UnpackSiblings(HashFunctionPoseidon, []byte{1, 0})
        c.Assert(err, qt.IsNotNil)
        _, err = UnpackSiblings(HashFunctionPoseidon, []byte{1, 0, 1, 0})
        c.Assert(err, qt.IsNotNil)
}
```

**Comment by altergui** [comment]
please make it `if len(b) < 4 {`
else there's still a panic possible:
```
	_, err = UnpackSiblings(HashFunctionPoseidon, []byte{3, 0, 0})
```

produces

```
--- FAIL: TestPackAndUnpackSiblings (0.00s)
panic: runtime error: slice bounds out of range [:4] with capacity 3 [recovered]
	panic: runtime error: slice bounds out of range [:4] with capacity 3

goroutine 25 [running]:
testing.tRunner.func1.2({0xc3ec40, 0xc00003ed38})
	/home/user/sdk/go1.20.5/src/testing/testing.go:1526 +0x24e
testing.tRunner.func1()
	/home/user/sdk/go1.20.5/src/testing/testing.go:1529 +0x39f
panic({0xc3ec40, 0xc00003ed38})
	/home/user/sdk/go1.20.5/src/runtime/panic.go:884 +0x213
go.vocdoni.io/dvote/tree/arbo.UnpackSiblings({0xef59d0, 0x1430310}, {0xc00028d4ea, 0x3, 0x3})
	/home/user/src/vocdoni/vocdoni-node/tree/arbo/tree.go:1022 +0x6a5
go.vocdoni.io/dvote/tree/arbo.TestPackAndUnpackSiblings(0xc00048bba0)
	/home/user/src/vocdoni/vocdoni-node/tree/arbo/tree_test.go:417 +0xbac
testing.tRunner(0xc00048bba0, 0xe3b5f8)
	/home/user/sdk/go1.20.5/src/testing/testing.go:1576 +0x10b
created by testing.(*T).Run
	/home/user/sdk/go1.20.5/src/testing/testing.go:1629 +0x3ea
FAIL	go.vocdoni.io/dvote/tree/arbo	0.047s
FAIL
```

**Comment by altergui** [comment]
well in fact the whole func is so brittle,
```
	_, err = UnpackSiblings(HashFunctionPoseidon, []byte{4, 0, 1, 0})
```
 also makes it panic
``` 
 --- FAIL: TestPackAndUnpackSiblings (0.00s)
panic: runtime error: slice bounds out of range [:5] with capacity 4 [recovered]
	panic: runtime error: slice bounds out of range [:5] with capacity 4

goroutine 35 [running]:
testing.tRunner.func1.2({0xc3ec40, 0xc0002901e0})
	/home/user/sdk/go1.20.5/src/testing/testing.go:1526 +0x24e
testing.tRunner.func1()
	/home/user/sdk/go1.20.5/src/testing/testing.go:1529 +0x39f
panic({0xc3ec40, 0xc0002901e0})
	/home/user/sdk/go1.20.5/src/runtime/panic.go:884 +0x213
go.vocdoni.io/dvote/tree/arbo.UnpackSiblings({0xef59d0, 0x1430310}, {0xc0002c5890, 0x4, 0x4})
	/home/user/src/vocdoni/vocdoni-node/tree/arbo/tree.go:1023 +0x693
go.vocdoni.io/dvote/tree/arbo.TestPackAndUnpackSiblings(0xc000503040)
	/home/user/src/vocdoni/vocdoni-node/tree/arbo/tree_test.go:420 +0xc85
testing.tRunner(0xc000503040, 0xe3b5f8)
	/home/user/sdk/go1.20.5/src/testing/testing.go:1576 +0x10b
created by testing.(*T).Run
	/home/user/sdk/go1.20.5/src/testing/testing.go:1629 +0x3ea
FAIL	go.vocdoni.io/dvote/tree/arbo	0.051s
FAIL
```

because of the `bitmapBytes := b[4 : 4+l]`

**Comment by altergui** [comment]
just for reference, this was fixed (squashed) as part of f40bf38b
